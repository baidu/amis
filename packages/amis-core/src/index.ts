export * from './abc';

export function hello() {
  return 'world';
}
