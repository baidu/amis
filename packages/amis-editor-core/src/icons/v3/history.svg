<?xml version="1.0" encoding="UTF-8"?>
<svg width="16px" height="16px" viewBox="0 0 16 16" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <g id="流程编辑器调整" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g id="补充icon" transform="translate(-214.000000, -23.000000)">
            <g id="历史记录" transform="translate(214.000000, 23.000000)">
                <rect id="矩形" x="0" y="0" width="16" height="16"></rect>
                <path d="M8,14.4978081 C11.5886403,14.4978081 14.4978081,11.5886403 14.4978081,8 C14.4978081,4.41135969 11.5886403,1.50219192 8,1.50219192 C6.61654784,1.50219192 5.33407683,1.93454364 4.28033323,2.67150085 C2.8531933,3.66960044 1.84560113,5.22642867 1.57491,7.02463225 C1.52701716,7.34278482 1.50219192,7.66849329 1.50219192,8 C1.50219192,11.5886403 4.41135969,14.4978081 8,14.4978081 Z" id="椭圆形" stroke="currentColor" stroke-linejoin="round"></path>
                <polyline id="路径-6" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" points="8 4.53234508 8 8.04122906 11.0455956 9.96470882"></polyline>
            </g>
        </g>
    </g>
</svg>