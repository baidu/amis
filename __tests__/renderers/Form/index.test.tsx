import React = require('react');
import PageRenderer from '../../../src/renderers/Form';
import * as renderer from 'react-test-renderer';
import {render, fireEvent, cleanup, getByText} from 'react-testing-library';
import '../../../src/themes/default';
import {render as amisRender} from '../../../src/index';
import {wait, makeEnv} from '../../helper';
import {clearStoresCache} from '../../../src/factory';
import {createMemoryHistory} from 'history';

// mock getComputedStyle
Object.defineProperty(window, 'getComputedStyle', {
  value: () => ({
    getPropertyValue: (prop: any) => {
      return '';
    }
  })
});

afterEach(() => {
  cleanup();
  clearStoresCache();
});

test('Renderer:Form', async () => {
  const resultPromise = Promise.resolve({
    data: {
      status: 0,
      msg: 'ok'
    }
  });
  const fetcher = jest.fn().mockImplementation(() => resultPromise);
  const {container, getByText} = render(
    amisRender(
      {
        type: 'form',
        api: '/api/xxx',
        controls: [
          {
            type: 'text',
            name: 'a',
            label: 'Label',
            value: '123'
          }
        ],
        title: 'The form',
        actions: [
          {
            type: 'submit',
            label: 'Submit'
          }
        ]
      },
      {},
      makeEnv({
        fetcher
      })
    )
  );
  expect(container).toMatchSnapshot();

  fireEvent.click(getByText('Submit'));
  await resultPromise;
  await wait(100);

  expect(fetcher).toHaveBeenCalled();
  expect(fetcher.mock.calls[0][0]).toMatchSnapshot();
});

test('Renderer:Form:valdiate', async () => {
  const notify = jest.fn();
  const onSubmit = jest.fn();
  const {container, getByText} = render(
    amisRender(
      {
        type: 'form',
        controls: [
          {
            type: 'text',
            name: 'a',
            required: true,
            label: 'Label'
          }
        ],
        title: 'The form',
        actions: [
          {
            type: 'submit',
            label: 'Submit'
          }
        ]
      },
      {
        onSubmit
      },
      makeEnv({
        notify
      })
    )
  );

  fireEvent.click(getByText('Submit'));
  expect(container).toMatchSnapshot();
  expect(onSubmit).not.toHaveBeenCalled();

  await wait(100);
  expect(notify).toHaveBeenCalledWith('error', '依赖的部分字段没有通过验证');

  const input = container.querySelector('input[name=a]');
  expect(input).toBeTruthy();
  fireEvent.change(input!, {
    target: {
      value: '123'
    }
  });
  await wait(300); // 有 250 秒左右的节流
  fireEvent.click(getByText('Submit'));
  expect(container).toMatchSnapshot();

  await wait(100);
  expect(onSubmit).toHaveBeenCalled();
  expect(onSubmit.mock.calls[0][0]).toMatchSnapshot();
});

test('Renderer:Form:remoteValidate', async () => {
  const notify = jest.fn();
  const fetcher = jest.fn().mockImplementation(() =>
    Promise.resolve({
      data: {
        status: 422,
        msg: '服务端验证失败',
        errors: {
          a: '这个字段服务端验证失败'
        }
      }
    })
  );
  const {container, getByText} = render(
    amisRender(
      {
        type: 'form',
        api: '/api/xxx',
        controls: [
          {
            type: 'text',
            name: 'a',
            label: 'Label'
          }
        ],
        title: 'The form',
        actions: [
          {
            type: 'submit',
            label: 'Submit'
          }
        ]
      },
      {},
      makeEnv({
        notify,
        fetcher
      })
    )
  );

  fireEvent.click(getByText('Submit'));
  await wait(100);
  expect(container).toMatchSnapshot();
});

test('Renderer:Form:onValidate', async () => {
  const notify = jest.fn();
  const onSubmit = jest.fn();
  const onValidate = jest
    .fn()
    .mockImplementationOnce(() => ({
      a: 'a is wrong',
      b: ['b is wrong', 'b is wrong 2']
    }))
    .mockImplementationOnce(() => ({
      a: '',
      b: ''
    }));
  const {container, getByText} = render(
    amisRender(
      {
        type: 'form',
        controls: [
          {
            type: 'text',
            name: 'a',
            label: 'A',
            value: 1
          },
          {
            type: 'text',
            name: 'b',
            label: 'B',
            value: 2
          }
        ],
        title: 'The form',
        actions: [
          {
            type: 'submit',
            label: 'Submit'
          }
        ]
      },
      {
        onSubmit,
        onValidate
      },
      makeEnv({
        notify
      })
    )
  );

  fireEvent.click(getByText('Submit'));
  await wait(100);

  expect(container).toMatchSnapshot();
  expect(onSubmit).not.toHaveBeenCalled();
  expect(onValidate).toHaveBeenCalled();
  expect(onValidate.mock.calls[0][0]).toMatchSnapshot();

  await wait(100);
  expect(notify).toHaveBeenCalledWith('error', '依赖的部分字段没有通过验证');

  fireEvent.click(getByText('Submit'));
  await wait(100);

  expect(container).toMatchSnapshot();
  expect(onSubmit).toHaveBeenCalled();
  expect(onSubmit.mock.calls[0][0]).toMatchSnapshot();
});

test('Renderer:Form initApi', async () => {
  const notify = jest.fn();
  let p0;
  const fetcher = jest.fn().mockImplementation(
    () =>
      (p0 = Promise.resolve({
        data: {
          status: 0,
          data: {
            a: 1,
            b: 2
          }
        }
      }))
  );
  const {container, getByText} = render(
    amisRender(
      {
        type: 'form',
        initApi: '/api/xxx',
        controls: [
          {
            type: 'text',
            name: 'a',
            label: 'A'
          },
          {
            type: 'text',
            name: 'b',
            label: 'B'
          }
        ],
        title: 'The form'
      },
      {},
      makeEnv({
        notify,
        fetcher
      })
    )
  );

  // fetch 调用了，所有 initApi 接口调用了
  expect(fetcher).toHaveBeenCalled();
  await p0;
  await wait(10);

  // 通过 snapshot 可断定 initApi 返回值已经作用到了表单项上。
  expect(container).toMatchSnapshot();
});

test('Renderer:Form initFetch:false', async () => {
  const notify = jest.fn();
  const fetcher = jest.fn().mockImplementation(() =>
    Promise.resolve({
      data: {
        status: 0,
        data: {
          a: 1,
          b: 2
        }
      }
    })
  );
  const {container, getByText} = render(
    amisRender(
      {
        type: 'form',
        initApi: '/api/xxx',
        initFetch: false,
        controls: [
          {
            type: 'text',
            name: 'a',
            label: 'A'
          },
          {
            type: 'text',
            name: 'b',
            label: 'B'
          }
        ],
        title: 'The form'
      },
      {},
      makeEnv({
        notify,
        fetcher
      })
    )
  );

  expect(fetcher).not.toHaveBeenCalled();
});

test('Renderer:Form initFetchOn:false', async () => {
  const notify = jest.fn();
  const fetcher = jest.fn().mockImplementation(() =>
    Promise.resolve({
      data: {
        status: 0,
        data: {
          a: 1,
          b: 2
        }
      }
    })
  );
  const {container, getByText} = render(
    amisRender(
      {
        type: 'form',
        initApi: '/api/xxx',
        initFetchOn: 'this.goFetch',
        controls: [
          {
            type: 'text',
            name: 'a',
            label: 'A'
          },
          {
            type: 'text',
            name: 'b',
            label: 'B'
          }
        ],
        title: 'The form'
      },
      {
        data: {
          goFetch: false
        }
      },
      makeEnv({
        notify,
        fetcher
      })
    )
  );

  expect(fetcher).not.toHaveBeenCalled();
});

test('Renderer:Form sendOn:false', async () => {
  const notify = jest.fn();
  const fetcher = jest.fn().mockImplementation(() =>
    Promise.resolve({
      data: {
        status: 0,
        data: {
          a: 1,
          b: 2
        }
      }
    })
  );
  const {container, getByText} = render(
    amisRender(
      {
        type: 'form',
        initApi: {
          method: 'get',
          url: '/api/xxx',
          sendOn: 'this.goFetch'
        },
        controls: [
          {
            type: 'text',
            name: 'a',
            label: 'A'
          },
          {
            type: 'text',
            name: 'b',
            label: 'B'
          }
        ],
        title: 'The form'
      },
      {
        data: {
          goFetch: false
        }
      },
      makeEnv({
        notify,
        fetcher
      })
    )
  );

  expect(fetcher).not.toHaveBeenCalled();
});

test('Renderer:Form sendOn:true', async () => {
  const notify = jest.fn();
  let p0;
  const fetcher = jest.fn().mockImplementation(
    () =>
      (p0 = Promise.resolve({
        data: {
          status: 0,
          data: {
            a: 1,
            b: 2
          }
        }
      }))
  );
  const {container, getByText} = render(
    amisRender(
      {
        type: 'form',
        initApi: {
          method: 'get',
          url: '/api/xxx',
          sendOn: 'this.goFetch'
        },
        controls: [
          {
            type: 'text',
            name: 'a',
            label: 'A'
          },
          {
            type: 'text',
            name: 'b',
            label: 'B'
          }
        ],
        title: 'The form'
      },
      {
        data: {
          goFetch: true
        }
      },
      makeEnv({
        notify,
        fetcher
      })
    )
  );

  expect(fetcher).toHaveBeenCalled();
  await p0;
  await wait(10);
  expect(container).toMatchSnapshot();
});
