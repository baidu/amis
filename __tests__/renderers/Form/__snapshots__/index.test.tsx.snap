// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer:Form 1`] = `
<div>
  <div
    class="a-Panel a-Panel--default a-Panel--form"
    style="position: relative;"
  >
    <div
      class="a-Panel-heading"
    >
      <h3
        class="a-Panel-title"
      >
        <span
          class="a-TplField"
        >
          <span>
            The form
          </span>
        </span>
      </h3>
    </div>
    <div
      class="a-Panel-body"
    >
      <form
        class="a-Form a-Form--normal"
        novalidate=""
      >
        <div
          class="a-Form-item a-Form-item--normal"
          data-role="form-item"
        >
          <label
            class="a-Form-label"
          >
            <span>
              <span
                class="a-TplField"
              >
                <span>
                  Label
                </span>
              </span>
            </span>
          </label>
          <div
            class="a-Form-control a-TextControl"
          >
            <div
              class="a-TextControl-input"
            >
              <input
                autocomplete="off"
                name="a"
                placeholder=""
                size="10"
                type="text"
                value="123"
              />
            </div>
          </div>
        </div>
        <input
          style="display: none;"
          type="submit"
        />
      </form>
    </div>
    <div
      class="a-Panel-footerWrap"
    >
      <div
        class="a-Panel-btnToolbar a-Panel-footer"
      >
        <button
          class="a-Button a-Button--default"
          type="submit"
        >
          <span>
            Submit
          </span>
        </button>
      </div>
    </div>
    <div
      class="resize-sensor"
      style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: scroll; z-index: -1; visibility: hidden;"
    >
      
  
      <div
        class="resize-sensor-expand"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0px; top: 0px; width: 10px; height: 10px;"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-shrink"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0; top: 0; width: 200%; height: 200%"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-appear"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;animation-name: apearSensor; animation-duration: 0.2s;"
      />
    </div>
  </div>
</div>
`;

exports[`Renderer:Form 2`] = `
Object {
  "body": Object {
    "a": "123",
  },
  "config": Object {
    "errorMessage": "saveFailed",
    "method": "post",
    "onSuccess": [Function],
    "successMessage": "saveSuccess",
  },
  "data": Object {
    "a": "123",
  },
  "method": "post",
  "url": "/api/xxx",
}
`;

exports[`Renderer:Form initApi 1`] = `
<div>
  <div
    class="a-Panel a-Panel--default a-Panel--form"
    style="position: relative;"
  >
    <div
      class="a-Panel-heading"
    >
      <h3
        class="a-Panel-title"
      >
        <span
          class="a-TplField"
        >
          <span>
            The form
          </span>
        </span>
      </h3>
    </div>
    <div
      class="a-Panel-body"
    >
      <form
        class="a-Form a-Form--normal"
        novalidate=""
      >
        <div
          class="a-Spinner-overlay"
        />
        <div
          class="a-Spinner a-Spinner--overlay"
        />
        <div
          class="a-Form-item a-Form-item--normal"
          data-role="form-item"
        >
          <label
            class="a-Form-label"
          >
            <span>
              <span
                class="a-TplField"
              >
                <span>
                  A
                </span>
              </span>
            </span>
          </label>
          <div
            class="a-Form-control a-TextControl"
          >
            <div
              class="a-TextControl-input"
            >
              <input
                autocomplete="off"
                name="a"
                placeholder=""
                size="10"
                type="text"
                value="1"
              />
            </div>
          </div>
        </div>
        <div
          class="a-Form-item a-Form-item--normal"
          data-role="form-item"
        >
          <label
            class="a-Form-label"
          >
            <span>
              <span
                class="a-TplField"
              >
                <span>
                  B
                </span>
              </span>
            </span>
          </label>
          <div
            class="a-Form-control a-TextControl"
          >
            <div
              class="a-TextControl-input"
            >
              <input
                autocomplete="off"
                name="b"
                placeholder=""
                size="10"
                type="text"
                value="2"
              />
            </div>
          </div>
        </div>
        <input
          style="display: none;"
          type="submit"
        />
      </form>
    </div>
    <div
      class="a-Panel-footerWrap"
    >
      <div
        class="a-Panel-btnToolbar a-Panel-footer"
      >
        <button
          class="a-Button a-Button--primary"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
    <div
      class="resize-sensor"
      style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: scroll; z-index: -1; visibility: hidden;"
    >
      
  
      <div
        class="resize-sensor-expand"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0px; top: 0px; width: 10px; height: 10px;"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-shrink"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0; top: 0; width: 200%; height: 200%"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-appear"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;animation-name: apearSensor; animation-duration: 0.2s;"
      />
    </div>
  </div>
</div>
`;

exports[`Renderer:Form sendOn:true 1`] = `
<div>
  <div
    class="a-Panel a-Panel--default a-Panel--form"
    style="position: relative;"
  >
    <div
      class="a-Panel-heading"
    >
      <h3
        class="a-Panel-title"
      >
        <span
          class="a-TplField"
        >
          <span>
            The form
          </span>
        </span>
      </h3>
    </div>
    <div
      class="a-Panel-body"
    >
      <form
        class="a-Form a-Form--normal"
        novalidate=""
      >
        <div
          class="a-Spinner-overlay"
        />
        <div
          class="a-Spinner a-Spinner--overlay"
        />
        <div
          class="a-Form-item a-Form-item--normal"
          data-role="form-item"
        >
          <label
            class="a-Form-label"
          >
            <span>
              <span
                class="a-TplField"
              >
                <span>
                  A
                </span>
              </span>
            </span>
          </label>
          <div
            class="a-Form-control a-TextControl"
          >
            <div
              class="a-TextControl-input"
            >
              <input
                autocomplete="off"
                name="a"
                placeholder=""
                size="10"
                type="text"
                value="1"
              />
            </div>
          </div>
        </div>
        <div
          class="a-Form-item a-Form-item--normal"
          data-role="form-item"
        >
          <label
            class="a-Form-label"
          >
            <span>
              <span
                class="a-TplField"
              >
                <span>
                  B
                </span>
              </span>
            </span>
          </label>
          <div
            class="a-Form-control a-TextControl"
          >
            <div
              class="a-TextControl-input"
            >
              <input
                autocomplete="off"
                name="b"
                placeholder=""
                size="10"
                type="text"
                value="2"
              />
            </div>
          </div>
        </div>
        <input
          style="display: none;"
          type="submit"
        />
      </form>
    </div>
    <div
      class="a-Panel-footerWrap"
    >
      <div
        class="a-Panel-btnToolbar a-Panel-footer"
      >
        <button
          class="a-Button a-Button--primary"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
    <div
      class="resize-sensor"
      style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: scroll; z-index: -1; visibility: hidden;"
    >
      
  
      <div
        class="resize-sensor-expand"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0px; top: 0px; width: 10px; height: 10px;"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-shrink"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0; top: 0; width: 200%; height: 200%"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-appear"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;animation-name: apearSensor; animation-duration: 0.2s;"
      />
    </div>
  </div>
</div>
`;

exports[`Renderer:Form:onValidate 1`] = `
<div>
  <div
    class="a-Panel a-Panel--default a-Panel--form"
    style="position: relative;"
  >
    <div
      class="a-Panel-heading"
    >
      <h3
        class="a-Panel-title"
      >
        <span
          class="a-TplField"
        >
          <span>
            The form
          </span>
        </span>
      </h3>
    </div>
    <div
      class="a-Panel-body"
    >
      <form
        class="a-Form a-Form--normal"
        novalidate=""
      >
        <div
          class="a-Form-item a-Form-item--normal is-error"
          data-role="form-item"
        >
          <label
            class="a-Form-label"
          >
            <span>
              <span
                class="a-TplField"
              >
                <span>
                  A
                </span>
              </span>
            </span>
          </label>
          <div
            class="a-Form-control is-error a-TextControl"
          >
            <div
              class="a-TextControl-input"
            >
              <input
                autocomplete="off"
                name="a"
                placeholder=""
                size="10"
                type="text"
                value="1"
              />
            </div>
          </div>
          <ul
            class="a-Form-feedback"
          >
            <li>
              a is wrong
            </li>
          </ul>
        </div>
        <div
          class="a-Form-item a-Form-item--normal is-error"
          data-role="form-item"
        >
          <label
            class="a-Form-label"
          >
            <span>
              <span
                class="a-TplField"
              >
                <span>
                  B
                </span>
              </span>
            </span>
          </label>
          <div
            class="a-Form-control is-error a-TextControl"
          >
            <div
              class="a-TextControl-input"
            >
              <input
                autocomplete="off"
                name="b"
                placeholder=""
                size="10"
                type="text"
                value="2"
              />
            </div>
          </div>
          <ul
            class="a-Form-feedback"
          >
            <li>
              b is wrong
            </li>
            <li>
              b is wrong 2
            </li>
          </ul>
        </div>
        <input
          style="display: none;"
          type="submit"
        />
      </form>
    </div>
    <div
      class="a-Panel-footerWrap"
    >
      <div
        class="a-Panel-btnToolbar a-Panel-footer"
      >
        <button
          class="a-Button a-Button--default"
          type="submit"
        >
          <span>
            Submit
          </span>
        </button>
      </div>
    </div>
    <div
      class="resize-sensor"
      style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: scroll; z-index: -1; visibility: hidden;"
    >
      
  
      <div
        class="resize-sensor-expand"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0px; top: 0px; width: 10px; height: 10px;"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-shrink"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0; top: 0; width: 200%; height: 200%"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-appear"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;animation-name: apearSensor; animation-duration: 0.2s;"
      />
    </div>
  </div>
</div>
`;

exports[`Renderer:Form:onValidate 2`] = `
Object {
  "a": 1,
  "b": 2,
}
`;

exports[`Renderer:Form:onValidate 3`] = `
<div>
  <div
    class="a-Panel a-Panel--default a-Panel--form"
    style="position: relative;"
  >
    <div
      class="a-Panel-heading"
    >
      <h3
        class="a-Panel-title"
      >
        <span
          class="a-TplField"
        >
          <span>
            The form
          </span>
        </span>
      </h3>
    </div>
    <div
      class="a-Panel-body"
    >
      <form
        class="a-Form a-Form--normal"
        novalidate=""
      >
        <div
          class="a-Form-item a-Form-item--normal"
          data-role="form-item"
        >
          <label
            class="a-Form-label"
          >
            <span>
              <span
                class="a-TplField"
              >
                <span>
                  A
                </span>
              </span>
            </span>
          </label>
          <div
            class="a-Form-control a-TextControl"
          >
            <div
              class="a-TextControl-input"
            >
              <input
                autocomplete="off"
                name="a"
                placeholder=""
                size="10"
                type="text"
                value="1"
              />
            </div>
          </div>
        </div>
        <div
          class="a-Form-item a-Form-item--normal"
          data-role="form-item"
        >
          <label
            class="a-Form-label"
          >
            <span>
              <span
                class="a-TplField"
              >
                <span>
                  B
                </span>
              </span>
            </span>
          </label>
          <div
            class="a-Form-control a-TextControl"
          >
            <div
              class="a-TextControl-input"
            >
              <input
                autocomplete="off"
                name="b"
                placeholder=""
                size="10"
                type="text"
                value="2"
              />
            </div>
          </div>
        </div>
        <input
          style="display: none;"
          type="submit"
        />
      </form>
    </div>
    <div
      class="a-Panel-footerWrap"
    >
      <div
        class="a-Panel-btnToolbar a-Panel-footer"
      >
        <button
          class="a-Button a-Button--default"
          type="submit"
        >
          <span>
            Submit
          </span>
        </button>
      </div>
    </div>
    <div
      class="resize-sensor"
      style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: scroll; z-index: -1; visibility: hidden;"
    >
      
  
      <div
        class="resize-sensor-expand"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0px; top: 0px; width: 10px; height: 10px;"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-shrink"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0; top: 0; width: 200%; height: 200%"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-appear"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;animation-name: apearSensor; animation-duration: 0.2s;"
      />
    </div>
  </div>
</div>
`;

exports[`Renderer:Form:onValidate 4`] = `
Object {
  "a": 1,
  "b": 2,
}
`;

exports[`Renderer:Form:remoteValidate 1`] = `
<div>
  <div
    class="a-Panel a-Panel--default a-Panel--form"
    style="position: relative;"
  >
    <div
      class="a-Panel-heading"
    >
      <h3
        class="a-Panel-title"
      >
        <span
          class="a-TplField"
        >
          <span>
            The form
          </span>
        </span>
      </h3>
    </div>
    <div
      class="a-Panel-body"
    >
      <form
        class="a-Form a-Form--normal"
        novalidate=""
      >
        <div
          class="a-Spinner-overlay"
        />
        <div
          class="a-Spinner a-Spinner--overlay"
        />
        <div
          class="a-Form-item a-Form-item--normal is-error"
          data-role="form-item"
        >
          <label
            class="a-Form-label"
          >
            <span>
              <span
                class="a-TplField"
              >
                <span>
                  Label
                </span>
              </span>
            </span>
          </label>
          <div
            class="a-Form-control is-error a-TextControl"
          >
            <div
              class="a-TextControl-input"
            >
              <input
                autocomplete="off"
                name="a"
                placeholder=""
                size="10"
                type="text"
                value=""
              />
            </div>
          </div>
          <ul
            class="a-Form-feedback"
          >
            <li>
              这个字段服务端验证失败
            </li>
          </ul>
        </div>
        <input
          style="display: none;"
          type="submit"
        />
      </form>
    </div>
    <div
      class="a-Panel-footerWrap"
    >
      <div
        class="a-Panel-btnToolbar a-Panel-footer"
      >
        <button
          class="a-Button a-Button--default"
          type="submit"
        >
          <span>
            Submit
          </span>
        </button>
      </div>
    </div>
    <div
      class="resize-sensor"
      style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: scroll; z-index: -1; visibility: hidden;"
    >
      
  
      <div
        class="resize-sensor-expand"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0px; top: 0px; width: 10px; height: 10px;"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-shrink"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0; top: 0; width: 200%; height: 200%"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-appear"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;animation-name: apearSensor; animation-duration: 0.2s;"
      />
    </div>
  </div>
</div>
`;

exports[`Renderer:Form:valdiate 1`] = `
<div>
  <div
    class="a-Panel a-Panel--default a-Panel--form"
    style="position: relative;"
  >
    <div
      class="a-Panel-heading"
    >
      <h3
        class="a-Panel-title"
      >
        <span
          class="a-TplField"
        >
          <span>
            The form
          </span>
        </span>
      </h3>
    </div>
    <div
      class="a-Panel-body"
    >
      <form
        class="a-Form a-Form--normal"
        novalidate=""
      >
        <div
          class="a-Form-item a-Form-item--normal is-error is-required"
          data-role="form-item"
        >
          <label
            class="a-Form-label"
          >
            <span>
              <span
                class="a-TplField"
              >
                <span>
                  Label
                </span>
              </span>
              <span
                class="a-Form-star"
              >
                *
              </span>
            </span>
          </label>
          <div
            class="a-Form-control is-error a-TextControl"
          >
            <div
              class="a-TextControl-input"
            >
              <input
                autocomplete="off"
                name="a"
                placeholder=""
                size="10"
                type="text"
                value=""
              />
            </div>
          </div>
          <ul
            class="a-Form-feedback"
          >
            <li>
              这是必填项
            </li>
          </ul>
        </div>
        <input
          style="display: none;"
          type="submit"
        />
      </form>
    </div>
    <div
      class="a-Panel-footerWrap"
    >
      <div
        class="a-Panel-btnToolbar a-Panel-footer"
      >
        <button
          class="a-Button a-Button--default is-disabled"
          disabled=""
          type="submit"
        >
          <span>
            Submit
          </span>
        </button>
      </div>
    </div>
    <div
      class="resize-sensor"
      style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: scroll; z-index: -1; visibility: hidden;"
    >
      
  
      <div
        class="resize-sensor-expand"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0px; top: 0px; width: 10px; height: 10px;"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-shrink"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0; top: 0; width: 200%; height: 200%"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-appear"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;animation-name: apearSensor; animation-duration: 0.2s;"
      />
    </div>
  </div>
</div>
`;

exports[`Renderer:Form:valdiate 2`] = `
<div>
  <div
    class="a-Panel a-Panel--default a-Panel--form"
    style="position: relative;"
  >
    <div
      class="a-Panel-heading"
    >
      <h3
        class="a-Panel-title"
      >
        <span
          class="a-TplField"
        >
          <span>
            The form
          </span>
        </span>
      </h3>
    </div>
    <div
      class="a-Panel-body"
    >
      <form
        class="a-Form a-Form--normal"
        novalidate=""
      >
        <div
          class="a-Form-item a-Form-item--normal is-required"
          data-role="form-item"
        >
          <label
            class="a-Form-label"
          >
            <span>
              <span
                class="a-TplField"
              >
                <span>
                  Label
                </span>
              </span>
              <span
                class="a-Form-star"
              >
                *
              </span>
            </span>
          </label>
          <div
            class="a-Form-control a-TextControl"
          >
            <div
              class="a-TextControl-input"
            >
              <input
                autocomplete="off"
                name="a"
                placeholder=""
                size="10"
                type="text"
                value="123"
              />
            </div>
          </div>
        </div>
        <input
          style="display: none;"
          type="submit"
        />
      </form>
    </div>
    <div
      class="a-Panel-footerWrap"
    >
      <div
        class="a-Panel-btnToolbar a-Panel-footer"
      >
        <button
          class="a-Button a-Button--default"
          type="submit"
        >
          <span>
            Submit
          </span>
        </button>
      </div>
    </div>
    <div
      class="resize-sensor"
      style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: scroll; z-index: -1; visibility: hidden;"
    >
      
  
      <div
        class="resize-sensor-expand"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0px; top: 0px; width: 10px; height: 10px;"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-shrink"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0; top: 0; width: 200%; height: 200%"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-appear"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;animation-name: apearSensor; animation-duration: 0.2s;"
      />
    </div>
  </div>
</div>
`;

exports[`Renderer:Form:valdiate 3`] = `
Object {
  "a": "123",
}
`;
